SOPs for plutus-core

michaelpj opened the pull request:
Things of note:
- Conformance tests included.
- I did still put some tests in `plutus-core`, since they also test typechecking and we don't have typechecking conformance tests yet.
- The CEK machine obviously needs careful attention.
- I refactored `LetFloatIn` in order to be able to share some logic with the new term constructors as they are n-ary.
- I have stopped short of diving into the Agda.

<!--
IMPORTANT: if you are an external contributor, make sure you have read the "External contributors" section of CONTRIBUTING.

Here are some checklists you may like to use. Use your judgement.

This is just a checklist, all the normative suggestions are covered in more detail in CONTRIBUTING.
-->
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reviewer requested

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs:
@zliu would appreciate attention here. Tests still pass and I added some for the new constructs, so I think it's okay.

zliu41 replied:
Looks good.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/QuickCheck/ShrinkTypes.hs:
```diff
@@ -249,6 +250,8 @@ shrinkKindAndType ctx (k0, ty) =
       where
         kArg = unsafeInferKind ctx arg
+    -- TODO: shrink the list rather than going to empty list?
+    TySOP  _ _tyls       -> [ (Type (), TySOP () []) ]
```
@effectfully wasn't sure how to do this

effectfully replied:
I'll fix it.

zliu41 commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:
```diff
@@ -696,6 +725,22 @@ enterComputeCek = computeCek
+          _              -> returnCek ctx $ VConstr i (toList done')
+    -- s , case _ (C0 ... CN, ρ) ◅ constr i V1 .. Vm  ↦  s , [_ V1 ... Vm] ; ρ ▻ Ci
+    returnCek (FrameCases env cs ctx) e = case e of
+        (VConstr i args) -> case cs ^? ix i of
```
Does it make sense to use `Vector` for `cs` for faster indexing?

michaelpj replied:
Ah yes. This needs a Note. In short, I spent a _long_ time trying out various accumulator types here in search of more performance. I started out with `Vector`/`MutableVector` as the "obvious" choice, at @effectfully 's prompting I tried `Seq` and also `List`/`DList`. To my great surprise, there was no real difference, and so I've gone with `List`/`DList` as it's the simplest.

zliu41 commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/Ck.hs:
```diff
@@ -58,6 +59,7 @@ data CkValue uni fun =
   | VLamAbs Name (Type TyName uni ()) (Term TyName Name uni fun ())
   | VIWrap (Type TyName uni ()) (Type TyName uni ()) (CkValue uni fun)
   | VBuiltin (Term TyName Name uni fun ()) (BuiltinRuntime (CkValue uni fun))
+  | VConstr (Type TyName uni ()) Int [CkValue uni fun]
```
nvm, it would obviously be much more expensive.

zliu41 commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/Ck.hs:
```diff
@@ -129,12 +134,17 @@ instance HasConstant (CkValue uni fun) where
     fromConstant = VCon
 
+-- CK not CEK
+-- substitution not environments
```
Did you intend to leave these comments?

michaelpj replied:
You know, I have no idea where those are from

zliu41 commented on plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:
```diff
@@ -219,9 +227,31 @@ mapFun f = go where
 newtype Binder name = Binder { unBinder :: name }
+
+{- Note [Constr tag type]
+Constructor tags are not dynamically created, they can only come from the program itself.
```
Since tags are always from the program, does it make sense to use Word16 or Word32 to save some space? Word32 is definitely sufficient, and Word16 should be, too - I can't imagine someone needing more than 65536 constructors.

michaelpj replied:
`Word64` is encoded like other integral types: using only as many bytes as are needed. So it won't save any space in the serialized program, and since it's a machine word I doubt we can actually save runtime memory by using a smaller type.

michaelpj replied:
I mentioned this in the comment.

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Check/Normal.hs:
```diff
@@ -37,6 +38,8 @@ check (Unwrap _ t)           = check t
 check (Apply _ t1 t2)        = check t1 >> check t2
 check (TyAbs _ _ _ t)        = check t
+check (Constr _ ty _ es)     = normalType ty >> traverse_ check es
+check (Case _ ty arg cs)     = normalType ty >> check arg >> traverse_ check cs
```
Yes, in the typechecker.

zliu41 replied:
Ok, I thought this is the type checker and it is checking each `c \in cs` independently

michaelpj commented on plutus-core/cost-model/data/cekMachineCosts.json:
```diff
@@ -6,6 +6,8 @@
     "cekForceCost"   : {"exBudgetCPU":   23000, "exBudgetMemory": 100},
     "cekApplyCost"   : {"exBudgetCPU":   23000, "exBudgetMemory": 100},
-    "cekBuiltinCost" : {"exBudgetCPU":   23000, "exBudgetMemory": 100}
+    "cekBuiltinCost" : {"exBudgetCPU":   23000, "exBudgetMemory": 100},
```
I _think_ we can just use them, but I'd appreciate a double check from @kwxm

kwxm replied:
I was just wondering about that.  We should try to do some measurements.

michaelpj replied:
I think it's very likely to be extremely similar cost, so I think we can do this for now at least.

zliu41 reviewed (approved):
Looks good, but there should be some tests verifying that `plcVersionsAvailableIn` works as expected, in other words, one cannot use Case or Constr in Plutus V1/V2. There doesn't seem to be such tests at the moment.

michaelpj commented:
Yep, good shout, I'll add that.

michaelpj commented:
Very good shout: the tests were broken, because the version check was in the wrong place. Not in a way that would have hurt us on-chain, I think, but annoyingly wrong. I fixed that, fixed the tests, and added some more.

effectfully commented on plutus-core/plutus-ir/src/PlutusIR/Transform/CaseReduce.hs:
```diff
@@ -0,0 +1,18 @@
+processTerm :: Term tyname name uni fun a -> Term tyname name uni fun a
+processTerm = \case
+    Case ann _ (Constr _ _ i args) cs | Just c <- cs ^? wix i -> mkIterApp ann c args
+    t                                                         -> t
```
Cool.

effectfully commented on plutus-core/untyped-plutus-core/test/Generators.hs:
```diff
@@ -59,6 +57,8 @@ compareTerm (Force _ t ) (Force _ t')         = compareTerm t t'
 compareTerm (Constant _ x) (Constant _ y)     = x == y
 compareTerm (Builtin _ bi) (Builtin _ bi')    = bi == bi'
+compareTerm (Constr _ i es) (Constr _ i' es') = i == i' && all (uncurry compareTerm) (zip es es')
+compareTerm (Case _ arg cs) (Case _ arg' cs') = compareTerm arg arg' && all (uncurry compareTerm) (zip cs cs')
```
Again.

effectfully commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Instance/Flat.hs:
```diff
@@ -57,7 +59,7 @@ tags and their used/available encoding possibilities.
 |------------------|-------------------|-----------|-------|------|-----------|
 | default builtins | encodeBuiltin     | 7         | 128   | 54   | 74        |
-| Terms            | encodeTerm        | 4         | 16    | 8    | 8         |
+| Terms            | encodeTerm        | 4         | 16    | 10   | 6         |
```
Are these comments actually helpful to anybody? It's a hassle to keep them up-to-date.

effectfully commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Type.hs:
```diff
@@ -80,6 +81,11 @@ data Term name uni fun ann
+    -- TODO: worry about overflow, maybe use an Integer
+    -- TODO: try spine-strict list or strict list or vector
+    -- See Note [Constr tag type]
+    | Constr !ann !Word64 ![Term name uni fun ann]
```
This should be a strict list I believe. Perhaps a custom one whose elements are always `Term`s. Gonna be a major pain in the ass though, since no `fmap`, `traverse` etc would work. Well, at least it should be a strict list.

michaelpj replied:
If we don't monomorphise it (and it's not clear that we gain anything from doing that, it's not like we can unpack Term) then we should still get `fmap` etc. There are many strict list types on Hackage even.

michaelpj replied:
TBH, I kind of want to fix this by having a "for-evaluation" term type anyway, then we can fix the name type etc which will actually let us unpack.

effectfully commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Type.hs:
```diff
@@ -80,6 +81,11 @@ data Term name uni fun ann
+    -- TODO: try spine-strict list or strict list or vector
+    -- See Note [Constr tag type]
+    | Constr !ann !Word64 ![Term name uni fun ann]
+    | Case !ann !(Term name uni fun ann) ![Term name uni fun ann]
```
This look like it should be a vector or something. At the very least a strict list.

effectfully commented on plutus-core/untyped-plutus-core/test/Evaluation/Golden/caseNonTag.plc.golden:
```diff
@@ -0,0 +1,3 @@
+(Left An error has occurred:  error:
+A non-constructor value was scrutinitzed in a case expression
+Caused by: (con integer 1))
```
Could there be any possible benefit from allowing `case` to scrutinize built-in enumeration types such as `Bool`, `List`, `Data` etc?

michaelpj replied:
I can't think of any, I hate it :p

effectfully commented on plutus-core/untyped-plutus-core/test/Evaluation/Golden/case4.plc.golden:
```diff
@@ -0,0 +1 @@
+(Right (con string "hello"))
```
It's super hard to review these. We should pretty-print the programs alongside what they evaluate/typecheck to.

effectfully reviewed (commented):
Gosh it's big. 

Looks good to me, except I didn't review the type checkers (will do) and the optimizations (will not do).

michaelpj commented on plutus-conformance/test-cases/uplc/evaluation/term/case/case-3/case-3.uplc:
```diff
@@ -0,0 +1,4 @@
+-- select first branch and do computation with the args
+(program 1.1.0
+    (case (constr 0 (con integer 3) (con integer 2)) (lam x (lam y [(builtin addInteger) x y])) (lam x (lam y [(builtin addInteger) x y])))
```
yes

michaelpj commented on plutus-conformance/test-cases/uplc/evaluation/term/case/case-4/case-4.uplc:
```diff
@@ -0,0 +1,4 @@
+-- select second branch and do computation with the args
+(program 1.1.0
+    (case (constr 1 (con integer 3) (con integer 2)) (lam x (lam y [(builtin addInteger) x y])) (lam x (lam y [(builtin subtractInteger) x y])))
```
Yep, I should add one.

michaelpj commented on plutus-conformance/test-cases/uplc/evaluation/term/constr/constr-5/constr-5.uplc:
```diff
@@ -0,0 +1,4 @@
+-- constr is strict in all its arguments
+(program 1.1.0
+    (constr 0 (con integer) (error))
+)
```
yes

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Check/Normal.hs:
```diff
@@ -48,6 +51,7 @@ normalType :: Type tyname uni ann -> Either (NormCheckError tyname name uni fun
 normalType (TyFun _ i o)       = normalType i >> normalType o
 normalType (TyForall _ _ _ ty) = normalType ty
 normalType (TyIFix _ pat arg)  = normalType pat >> normalType arg
+normalType (TySOP _ tyls)      = (traverse_ . traverse_) normalType tyls
```
Sure.

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Core/Instance/Eq.hs:
```diff
@@ -89,13 +91,21 @@ eqTypeM (TyFun ann1 dom1 cod1) (TyFun ann2 dom2 cod2) = do
     eqM ann1 ann2
     eqM bi1 bi2
+eqTypeM (TySOP ann1 tyls1) (TySOP ann2 tyls2) = do
+    eqM ann1 ann2
+    case zipExact tyls1 tyls2 of
+        Just ps -> for_ ps $ \(ptys1, ptys2) -> case zipExact ptys1 ptys2 of
+          Just tys -> for_ tys $ \(ty1, ty2) -> eqTypeM ty1 ty2
+          Nothing  -> empty
+        Nothing -> empty
```
Huh, I didn't realise we could `lift` into it. I guess it depends on whether we want to "expose" the fact that there's a maybe in the monad stack :shrug: either seems fine to me

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Parser/Type.hs:
```diff
@@ -46,9 +46,15 @@ ifixType = withSpan $ \sp ->
+sopType = withSpan $ \sp -> inParens $ TySOP sp <$> (symbol "sop" *> many tyList)
+  where
+    tyList :: Parser [PType]
+    tyList = (inBrackets $ many pType) <* whitespace
```
I think this is @zliu41 . The whitespace handling became complicated once we started tracking source spans (this could do with a note TBH). Now we have to be careful about consuming trailing whitespace otherwise it will make the end of the source span look wrong. The correct solution would be a lexer, I think...

effectfully replied:
> Now we have to be careful about consuming trailing whitespace otherwise it will make the end of the source span look wrong.

Ah, of course, thanks.

mjaskelioff commented:
I realised there is a case missing in the CEK machine, the case for the frame  of left application to value.
$$[ \textunderscore V_1 ] \cdot s ~ \triangleleft V_2$$

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -372,6 +374,10 @@ inferKindM (TyIFix ann pat arg)    = do
     checkKindM ann pat $ toPatFuncKind k
     pure $ Type ()
 
+inferKindM (TySOP ann tyls)        = do
```
This one doesn't have a comment with a rule.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
     normalizeTypeM $ void ty
 
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_i ... s_n     s_i = [p_0 ... p_m]
+-- [check| G !- t_j : p_j]
```
I think I'd write the last line as

```
-- [check| G !- t_0 : p_0] ... [check| G !- t_m : p_m]
```

but whatever.

michaelpj replied:
It's awkward, I'm eliding a lot of iteration. If we start doing that then we should also include the iteration over the sums...

michaelpj replied:
I think I figured out a way to write it okay.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
+
+    -- We don't know exactly what to expect, we only know what the i-th sum should look like, so we
+    -- assert that we should have some types in the sum up to there, and then the known product type.
+    let expectedType = TySOP () (replicate ((fromIntegral i) - 1) [dummyType] ++ [replicate (length args) dummyType])
```
I don't get the comment. They all seem to be `dummyType` in there?

michaelpj replied:
It's unclear. I mean we know the _shape_ of one of the product types, not the whole thing.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
+
+    pure vResTy
+
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_n     s_i = [p_i_0 ... p_i_m]
```
Make it `sop s_0 ... s_i ... s_n` instead of `sop s_0 ... s_n` like in the rule above?
It's also `[p_i_0 ... p_i_m]` here and `[p_0 ... p_m]` there. I think I prefer the latter, since it has less clutter and it's easy to figure out the implicit `i` anyway.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
+    pure vResTy
+
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_n     s_i = [p_i_0 ... p_i_m]
+-- [check| G !- c_j : p_i_0 -> ... -> p_i_m -> vResTy]
```
The rule seems to be completely off? There's no mention of the scrutinee and the result is specified to be a SOP, which isn't the case in the actual logic.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
+    -- If we got through all that, then every case type is correct, including that
+    -- they all result in vResTy, so we can safely conclude that that is the type of the
+    -- whole expression.
+    pure vResTy
```
Good comment.

effectfully commented on plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs:
```diff
@@ -218,6 +221,58 @@ inferTypeM (Unwrap ann term)        = do
+-- [check| G !- t_j : p_j]
+-- ----------------------------------------------------------------------------
+-- [infer| G !- constr resTy i t_0 ... t_n : vResTy]
+inferTypeM t@(Constr ann resTy i args) = do
```
Is there really no way we could abstract this and the other one into a reusable function? Copy-pasting so much code feels very unsettling.

effectfully reviewed (commented):
Review of the type checkers. Mostly nitpicking, however I believe one rule written in a comment is wrong. The code looks good to me, except I hate the `dummyType` business.

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
+
+    -- We don't know exactly what to expect, we only know what the i-th sum should look like, so we
+    -- assert that we should have some types in the sum up to there, and then the known product type.
+    let expectedType = TySOP () (replicate ((fromIntegral i) - 1) [dummyType] ++ [replicate (length args) dummyType])
```
https://input-output.atlassian.net/browse/PLT-5837

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Core/Instance/Pretty/Classic.hs:
```diff
@@ -54,6 +54,10 @@ instance (PrettyClassicBy configName tyname, Pretty (SomeTypeIn uni), Pretty ann
         TyLam ann n k t    ->
             sexp "lam" (consAnnIf config ann
                 [prettyBy config n, prettyBy config k, prettyBy config t])
+        TySOP ann tyls    ->
+            sexp "sop" (consAnnIf config ann (fmap prettyTyl tyls))
+            where
+              prettyTyl tyl = brackets (sep (fmap (prettyBy config) tyl))
```
Yes. Perhaps a bad choice? But I chose to do it like application, i.e. `[a b c]`.

effectfully replied:
I just looked at things like

```
plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs:SOPs: @mkPatternFunctorBody List = sop [] [a, List a]@
plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs:        = fix (\(List :: * -> *) (a :: *) -> \(a :: *) -> sop [] [a, List a])
```

and was confused about the syntax. I don't mind separating types by spaces, but only because the classic syntax was doomed from the start :slightly_smiling_face:

michaelpj replied:
That's just me being inconsistent :upside_down_face:
